### PR TITLE
fix: Cache getJSONBody and getFormData response (Fixes #746)

### LINKED ISSUE #746: TypeError during Password Reset Flow in Next.js using SuperTokens
### Brief
I encountered a TypeError when attempting to reset a user's password in a Next.js application using SuperTokens. The error occurs when the user clicks "Change Password" after entering the new password and confirming it.

### Steps to Reproduce:
1. Create a new next.js application: `npx create-supertokens-app@latest`, Next.js -> App Directorty -> Social Login + Email Password
2. Run it locally
3. Create an account using email + password
4. Sign out
5. Reset password with the associated email
6. Open reset password link from the email received
7. Proceed to input new and confirm passwords
8. Click "Change password"
9. Observe the UI with an error "Something went wrong. Please try again."
10. Observe the terminal logged with `TypeError Body is unusable...`

Error sample:
```
TypeError: Body is unusable
at specConsumeBody (node:internal/deps/undici/undici:6630:15)
at NextRequest.json (node:internal/deps/undici/undici:6533:18)
at Object.getJSONBody (webpack-internal:///(rsc)/./node_modules/.pnpm/supertokens-node@16.5.1/node_modules/supertokens-node/lib/build/nextjs.js:71:38)
at PreParsedRequest.getJSONBody (webpack-internal:///(rsc)/./node_modules/.pnpm/supertokens-node@16.5.1/node_modules/supertokens-node/lib/build/framework/custom/framework.js:45:33)
at Object.passwordReset [as default] (webpack-internal:///(rsc)/./node_modules/.pnpm/supertokens-node@16.5.1/node_modules/supertokens-node/lib/build/recipe/emailpassword/api/passwordReset.js:36:36)
at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
at async Recipe.handleAPIRequest (webpack-internal:///(rsc)/./node_modules/.pnpm/supertokens-node@16.5.1/node_modules/supertokens-node/lib/build/recipe/emailpassword/recipe.js:95:24)
at async Recipe.handleAPIRequest (webpack-internal:///(rsc)/./node_modules/.pnpm/supertokens-node@16.5.1/node_modules/supertokens-node/lib/build/recipe/thirdpartyemailpassword/recipe.js:49:24)
at async SuperTokens.middleware (webpack-internal:///(rsc)/./node_modules/.pnpm/supertokens-node@16.5.1/node_modules/supertokens-node/lib/build/supertokens.js:160:38)
at async eval (webpack-internal:///(rsc)/./node_modules/.pnpm/supertokens-node@16.5.1/node_modules/supertokens-node/lib/build/framework/custom/framework.js:134:28)
at async handleCall (webpack-internal:///(rsc)/./node_modules/.pnpm/supertokens-node@16.5.1/node_modules/supertokens-node/lib/build/nextjs.js:76:40)
at async /Users/path-to-project/node_modules/.pnpm/next@14.0.3_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-route.runtime.dev.js:6:62609```

### PR DESCRIPTION
**Pull request by anku255**
## Summary of change

In many frameworks, calling `req.body` or `req.formData` is limited to a single use due to their nature as readable streams. Our framework classes have `getJSONBody` and `getFormData` functions, designed to invoke the appropriate method on the `req` object.

Calling these functions multiple times can lead to issues, as observed in #746. For instance, the `getJSONBody` was mistakenly called twice in the passwordReset API implementation. Although combining these calls into one resolved the specific case in #746, it does not prevent users from inadvertently triggering the same error by calling `getJSONBody` multiple times.

To address this, we now cache the response of `req.body` or `req.formData` directly in the recipe implementation.

## Related issues

-   https://github.com/supertokens/supertokens-node/issues/746

## Test Plan

Tests were added to ensure that `getJSONFromRequestBody` and `getFormDataFromRequestBody` are invoked only once, regardless of the number of calls to `getJSONData` or `getFormData`. If any framework class overrides the implementation of `getJSONData` or `getFormData` from the `BaseRequest` class, these tests would fail.

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [x] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/configUtils.ts` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

### CODE REVIEW

File: lib/ts/recipe/emailpassword/api/passwordReset.ts [hunk 1]
```diff
@@ -39,11 +41,11 @@ export default async function passwordReset(
         value: string;
     }[] = await validateFormFieldsOrThrowError(
         options.config.resetPasswordUsingTokenFeature.formFieldsForPasswordResetForm,
-        (await options.req.getJSONBody()).formFields,
+        requestBody.formFields,
         tenantId
     );
 
-    let token = (await options.req.getJSONBody()).token;
+    let token = requestBody.token;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/api/passwordReset.ts:48]: you should keep the old code here where we are reading it twice, and add a test for the froamework (at least for express and nextjs) where we cal this API and make sure it doesn't fail. You should check the test by removing the caching mechanism and manually making sure that the test does fail, and then readd the caching logic to make sure the test then passes.

Comment (rishabhpoddar) [reply lib/ts/recipe/emailpassword/api/passwordReset.ts:48]: Or if you don't like that, then have some other way to add tests for this PR.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: test/framework/awsLambda.test.js [hunk 2]
```diff
@@ -35,6 +35,8 @@ let Dashboard = require("../../recipe/dashboard");
 let { createUsers } = require("../utils");
 const { Querier } = require("../../lib/build/querier");
 const { maxVersion } = require("../../lib/build/utils");
+const { AWSRequest } = require("../../lib/build/framework/awsLambda/framework");
+const sinon = require("sinon");
```
Comment (rishabhpoddar) [inline test/framework/awsLambda.test.js:39]: if this is a new dependency, you need to add it in dev dependcy in package.json

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/framework/custom/framework.ts [hunk 3]
```diff
@@ -51,9 +50,13 @@ export class PreParsedRequest extends BaseRequest {
         this.request = request;
     }
 
-    getFormData = async (): Promise<any> => {
+    protected getJSONFromRequestBody(): Promise<any> {
```
[the remark this answers is not in the document]
Comment (anku255) [reply lib/ts/framework/custom/framework.ts:53]: ah.. right. I will fix it.

File: test/framework/awsLambda.test.js (hunk 2 above)
Comment (anku255) [reply test/framework/awsLambda.test.js:39]: No, `sinon` was already added as devDependency.

**Comment by rishabhpoddar** [comment]
Also, the AWS lambda PR check is failing.

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**